utils/create_merge_ids.py

```diff
@@ -114,82 +89,131 @@ def create_match_sets(
-                            max_merg = merg_id
-        ignore_ids = set()
-        for fi in os.listdir(ids_to_drop):
-            with open(os.path.join(ids_to_drop, fi)) as f:
+    prev_orig_to_merg = {}
+    merg_to_orig = {}
+    max_merg = "carticle_0"
+    if prev_id_mapping_dir is not None:
+        for fi in os.listdir(prev_id_mapping_dir):
+            with open(os.path.join(prev_id_mapping_dir, fi)) as f:
                 for line in f:
                     js = json.loads(line.strip())
-                    ignore_ids.add(js["merged_id"])
-        match_id = int(max_merg.split("carticle_")[1]) + 1
-        num_new, num_old = 0, 0
-        for ms in match_sets:
-            cset_article_id = None
-            # if we have exactly one existing id, reuse it, even if new articles are matched to it.
-            # if two articles that previously had different carticle ids are now in the same match set,
-            # create a new carticle id
-            existing_ids = set(
-                [prev_orig_to_merg[m] for m in ms if m in prev_orig_to_merg]
-            )
-            if len(existing_ids) == 1 and list(existing_ids)[0] not in ignore_ids:
-                cset_article_id = existing_ids.pop()
-                num_old += 1
+                    orig_id = js["orig_id"]
+                    merg_id = js["merged_id"]
+                    assert orig_id not in prev_orig_to_merg
+                    prev_orig_to_merg[orig_id] = merg_id
+                    if merg_id not in merg_to_orig:
+                        merg_to_orig[merg_id] = set()
+                    merg_to_orig[merg_id].add(orig_id)
+                    if merg_id > max_merg:
+                        max_merg = merg_id
+    ignore_ids = set()
+    for fi in os.listdir(ids_to_drop):
+        with open(os.path.join(ids_to_drop, fi)) as f:
+            for line in f:
+                js = json.loads(line.strip())
+                ignore_ids.add(js["merged_id"])
+    match_id = int(max_merg.split("carticle_")[1]) + 1
+    batch_size = 1_000_000
+    batch_count = 0
+    batch = []
+    for ms in match_sets:
+        cset_article_id = None
+        # if we have exactly one existing id, reuse it, even if new articles are matched to it.
+        # if two articles that previously had different carticle ids are now in the same match set,
+        # create a new carticle id
+        existing_ids = set([prev_orig_to_merg[m] for m in ms if m in prev_orig_to_merg])
+        if len(existing_ids) == 1 and list(existing_ids)[0] not in ignore_ids:
+            cset_article_id = existing_ids.pop()
+        # In some cases, merged ids can "split apart", if their constituent articles no longer
+        # match. We'll detect this case by checking whether the old set of articles assigned to
+        # this merged id contain any entries missing from our current set
+        if (cset_article_id and (len(merg_to_orig[cset_article_id] - set(ms)) > 0)) or (
```

Comment:
This is the one really salient code change - splitting up merged ids if they lose orig ids